src/diagnostics/MOM_harmonic_analysis.F90

```diff
@@ -0,0 +1,411 @@
+    call MOM_error(NOTE, trim(mesg))
+  endif
+
+end subroutine HA_accum
+
+!> This subroutine computes the harmonic constants and write output for the current field
+subroutine HA_write(ha1, Time, G, CS)
+  type(HA_type), pointer, intent(in) :: ha1
+  type(time_type),        intent(in) :: Time       !< The current model time
+  type(ocean_grid_type),  intent(in) :: G          !< The ocean's grid structure
+  type(HA_CS),            intent(in) :: CS         !< Control structure of the MOM_harmonic_analysis module
+
+  ! Local variables
+  real, dimension(:,:,:), allocatable :: FtSSHw    !< An array containing the harmonic constants
+  integer :: year, month, day, hour, minute, second
+  integer :: nc, k, is, ie, js, je
+
+  character(len=255)           :: filename         !< Output file name
+  type(MOM_infra_file)         :: cdf              !< The file handle for output harmonic constants
+  type(vardesc),   allocatable :: cdf_vars(:)      !< Output variable names
+  type(MOM_field), allocatable :: cdf_fields(:)    !< Field type variables for the output fields
+
+  nc = CS%nc ; is = ha1%is ; ie = ha1%ie ; js = ha1%js ; je = ha1%je
+
+  allocate(FtSSHw(is:ie,js:je,2*nc+1), source=0.0)
+
+  ! Compute the harmonic coefficients
+  call HA_solver(ha1, nc, FtSSHw)
+
+  ! Output file name
+  call get_date(Time, year, month, day, hour, minute, second)
+  write(filename, '(a,"HA_",a,i0.4,i0.2,i0.2,".nc")') &
+      trim(CS%path), trim(ha1%key), year, month, day
+
+  allocate(cdf_vars(2*nc+1))
+  allocate(cdf_fields(2*nc+1))
+
+  ! Variable names
+  cdf_vars(1) = var_desc("z0", "m" ,"mean value", 'h', '1')
+  do k=1,nc
+    cdf_vars(2*k  ) = var_desc(trim(CS%const_name(k))//"cos", "m", "cosine coefficient", 'h', '1')
+    cdf_vars(2*k+1) = var_desc(trim(CS%const_name(k))//"sin", "m", "sine coefficient",   'h', '1')
+  enddo
+
+  ! Create output file
+  call create_MOM_file(cdf, trim(filename), cdf_vars, &
+                       2*nc+1, cdf_fields, SINGLE_FILE, 86400.0, G=G)
+
+  ! Write data
+  call MOM_write_field(cdf, cdf_fields(1), G%domain, FtSSHw(:,:,1), 0.0)
+  do k=1,nc
+    call MOM_write_field(cdf, cdf_fields(2*k  ), G%domain, FtSSHw(:,:,2*k  ), 0.0)
+    call MOM_write_field(cdf, cdf_fields(2*k+1), G%domain, FtSSHw(:,:,2*k+1), 0.0)
+  enddo
+
+  call cdf%flush()
+  deallocate(cdf_vars)
+  deallocate(cdf_fields)
+
+end subroutine HA_write
+
+!> This subroutine computes the harmonic constants (stored in FtSSHw) using the dot products of the temporal
+!! basis functions accumulated in FtF, and the dot products of the SSH (or other fields) with the temporal basis
+!! functions accumulated in FtSSH. The system is solved by Cholesky decomposition.
+subroutine HA_solver(ha1, nc, FtSSHw)
+  type(HA_type), pointer,              intent(in)  :: ha1
+  integer,                             intent(in)  :: nc
+  real, dimension(:,:,:), allocatable, intent(out) :: FtSSHw !< Work array for Cholesky decomposition
```

Comment:
Please describe the units of FtSSHw, tmp and FtFw.  In this case, these might be arbitrary units (which we sometimes describe with `[A]`, but if they all have the same arbitrary units, that is useful to note, and all would use the same `[A]` descriptor.